File: source/sdk/wardnet-js/src/types/jobs.ts
/** Discriminator for what kind of background work a job represents. */
export type JobKind = "blocklist_refresh";

/** Lifecycle state of a background job. */
export type JobStatus = "PENDING" | "RUNNING" | "SUCCEED" | "TERMINATED_WITH_ERRORS";

/** Snapshot of a background job, returned by `GET /api/jobs/:id`. */
export interface Job {
  id: string;
  kind: JobKind;
  status: JobStatus;
  /** Completion percentage, 0..=100. */
  percentage_done: number;
  /** Populated when `status == "TERMINATED_WITH_ERRORS"`. */
  error: string | null;
  created_at: string;
  updated_at: string;
}

/** Response body returned when a handler dispatches a background job
 *  instead of performing the work inline. Status code is `202 Accepted`. */
export interface JobDispatchedResponse {
  job_id: string;
}

/** True when the job has reached a terminal state. Pollers should stop once
 *  they observe this — subsequent calls may return 404 after the GC TTL. */
export function isTerminal(status: JobStatus): boolean {
  return status === "SUCCEED" || status === "TERMINATED_WITH_ERRORS";
}

File: source/daemon/crates/wardnet-common/src/jobs.rs
//! Background-job types.
//!
//! Handlers that kick off work too slow for a synchronous HTTP response
//! dispatch a [`Job`] and return its id. The client polls `GET /api/jobs/:id`
//! until the job reaches a terminal state.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Discriminator for what kind of work a job represents. The kind is fixed
/// at job creation time; new kinds get new variants here.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JobKind {
    BlocklistRefresh,
}

/// Lifecycle state of a [`Job`]. Jobs start as `Pending`, move to `Running`
/// when execution begins, and end in one of the two terminal states.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum JobStatus {
    Pending,
    Running,
    #[serde(rename = "SUCCEED")]
    Succeeded,
    #[serde(rename = "TERMINATED_WITH_ERRORS")]
    Failed,
}

impl JobStatus {
    /// A job in a terminal state stays in the registry for the GC TTL and
    /// then disappears. Pollers should stop once they observe a terminal
    /// status.
    #[must_use]
    pub fn is_terminal(self) -> bool {
        matches!(self, JobStatus::Succeeded | JobStatus::Failed)
    }
}

/// Snapshot of a job, returned by `GET /api/jobs/:id`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Job {
    pub id: Uuid,
    pub kind: JobKind,
    pub status: JobStatus,
    /// Completion percentage, 0..=100. Jumps to 100 when the job terminates.
    pub percentage_done: u8,
    /// Populated when `status == Failed`.
    pub error: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Response body for a handler that dispatches a background job instead of
/// doing the work inline. The HTTP status code is `202 Accepted`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JobDispatchedResponse {
    pub job_id: Uuid,
}
